netlogo-gui/src/main/ide/CodeCompletionPopup.scala

```diff
@@ -0,0 +1,152 @@
+              println("here")
+              if (position != 0) {
+                eA.setCaretPosition(position - 1)
+                val doc = eA.getDocument.asInstanceOf[javax.swing.text.PlainDocument]
+                doc.remove(position - 1, 1)
+              }
+            case java.awt.event.KeyEvent.VK_ENTER =>
+              var suggestion = suggestionDisplaylist.getSelectedValue
+              lastSuggested = suggestion
+              val token = compiler.getTokenAtPosition(eA.getText(), eA.getCaretPosition)
+              val position = eA.getCaretPosition
+              val doc = eA.getDocument.asInstanceOf[javax.swing.text.PlainDocument]
+              doc.replace(token.start, position - token.start, suggestion, null)
+              eA.setCaretPosition(token.start + suggestion.length)
+              doc.removeDocumentListener(autoSuggestDocumentListener)
+              isPopupEnabled = false
+            case _ =>
+              e.setSource(eA)
+              eA.dispatchEvent(e)
+          }
+        }
+
+        override def keyPressed(e: KeyEvent): Unit = {}
+
+        override def keyReleased(e: KeyEvent): Unit = {
+        }
+      })
+    }
+  }
+
+  /**
+    * Makes the suggestion box initially and displays it.
+    * Should only be called when the user presses the shortcut to
+    * enable the suggestion. To update the suggestions list call showPopUp()
+    */
+  def displayPopup(): Unit = {
+    for{eA <- editorArea} {
+      isPopupEnabled = true
+      val token = compiler.getTokenAtPosition(eA.getText(), eA.getCaretPosition)
+      val position = if (token != null) token.start else eA.getCaretPosition
+      window.setLocation(eA.getLocationOnScreen.x + eA.modelToView(position).x,
+        eA.getLocationOnScreen.y + eA.modelToView(position).y + eA.getFont.getSize)
+      updatePopup()
+    }
+  }
+
+  /**
+    * Updates the list of the suggestion box and makes it visible.
+    */
+  def updatePopup(): Unit = {
+    println("I was called")
+    for{eA <- editorArea} {
+      if (isPopupEnabled) {
+        val token = compiler.getTokenAtPosition(eA.getText(), eA.getCaretPosition)
+        var list = Seq[String]()
+        // word only contains the current token till the cursor position
+        val word = if (token != null) token.text.substring(0, eA.getCaretPosition - token.start) else null
+        // popup not to be diplayed after user hits the enter key
+        if (word != null && !token.text.equals(lastSuggested)) {
+          list = autoSuggest.getSuggestions(word)
+          dlm.removeAllElements()
+          list.foreach(dlm.addElement(_))
+          if (!list.isEmpty) {
+            lastSuggested = ""
+            window.validate()
+            window.setVisible(true)
+            // Required to keep the caret in the editorArea visible
+            eA.getCaret.setVisible(true)
+            suggestionDisplaylist.setSelectedIndex(0)
+          } else {
+            isPopupEnabled = false
```

Comment:
It looks to me like there's a state machine made up of `isPopupEnabled`, `isVisible` and whether `autoSuggestListener` is registered. Think about whether that can be extracted into an object that manages the state to keep things simpler in the swing code.